tools/gcs-usage-recommender/python/backfill.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2019 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import json
+import sys
+import logging
+from typing import Dict, List
+from google.cloud import resource_manager
+from google.cloud import storage
+
+
+def get_storage_client() -> storage.Client:
+    """Retrieves GCS client.
+
+    Returns:
+        storage.Client object
+    """
+    return storage.Client()
+
+
+def get_resource_manager_client() -> resource_manager.Client:
+    """Retrieves Resource Manager Cleint
+
+    Returns:
+        resource_manager.Client object
+    """
+    return resource_manager.Client()
+
+
+def get_project_ids(resource_mgr_client: resource_manager.Client,
+                    organization_id: int) -> List[str]:
+    """Generates list of project_ids within a GCP org.
+
+    Args:
+        resource_mgr_client: resource_manager.Client
+        organization_id: Integer representing GCP org ID
+
+    Returns:
+        List of strings holding all project IDs.
+    """
+    project_id_list = []
+    project_filter = {
+        'parent.id': organization_id,
+        'parent.type': 'organization'
+    }
+    for project in resource_mgr_client.list_projects(project_filter):
+        project_id_list.append(project.project_id)
+    return project_id_list
+
+
+def get_buckets(project_ids: List[str],
+                gcs_client: storage.Client) -> List[Dict[str, str]]:
+    """Retrieves list of metadata for all buckets in a GCP org.
+
+    Args:
+        project_ids: List of strings holding project IDs
+        gcs_client: storage.Client object
+
+    Returns:
+        List of dictionaries mapping bucket-level metadata.
+    """
+    output_list = []
+    try:
+        for project_id in project_ids:
+            bucket_list = list(gcs_client.list_buckets(project=project_id))
```

Comment:
Hmm.. so imap is only available in Python 2, and I wanted to be conducive to both Python 2/3 environments. I also ran this in our PSO org, and it didn't seem to have performance issues. But LMK what you think @jaketf

Reply:
ah TIL imap in python2 -> map in python3 ([source](https://stackoverflow.com/questions/30271712/i-cant-find-imap-in-itertools-in-python)), thanks!
I think we should just leave this as is.

I think the way to do this preserving the iterator would be to use `map` and `chain.from_iterable` but it becomes much less readable and based on some testing we'd have to scale tot 100s of millions of buckets before there'd be real danger of memory issues on cloud shell / n1-standard1. I think my bias for iterator comes from object listing scenarios where millions of objects is more likely and where the benefit of lazily evaluated iterators can be safer.

If you're curious this would work in python2/3 and benefit from iterator if python3.
```python3
bucket_itr = itertools.chain.from_iterable(
    map(lambda project_id: gcs_client.list_buckets(project=project_id), 
             project_ids))
```